File: app/src/main/java/org/wikipedia/history/HistoryFragment.java
package org.wikipedia.history;

import android.app.AlertDialog;
import android.content.Context;
import android.database.Cursor;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.BackPressedHandler;
import org.wikipedia.Constants;
import org.wikipedia.R;
import org.wikipedia.WikipediaApp;
import org.wikipedia.activity.FragmentUtil;
import org.wikipedia.database.DatabaseClient;
import org.wikipedia.database.contract.PageHistoryContract;
import org.wikipedia.main.MainActivity;
import org.wikipedia.main.MainFragment;
import org.wikipedia.page.PageAvailableOfflineHandler;
import org.wikipedia.readinglist.database.ReadingList;
import org.wikipedia.settings.Prefs;
import org.wikipedia.util.FeedbackUtil;
import org.wikipedia.util.ResourceUtil;
import org.wikipedia.util.log.L;
import org.wikipedia.views.DefaultViewHolder;
import org.wikipedia.views.PageItemView;
import org.wikipedia.views.SearchEmptyView;
import org.wikipedia.views.SwipeableItemTouchHelperCallback;
import org.wikipedia.views.WikiCardView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class HistoryFragment extends Fragment implements BackPressedHandler {
    public interface Callback {
        void onLoadPage(@NonNull HistoryEntry entry);
        void onClearHistory();
    }

    private Unbinder unbinder;
    @BindView(R.id.history_list) RecyclerView historyList;
    @BindView(R.id.history_empty_container) View historyEmptyView;
    @BindView(R.id.search_empty_view) SearchEmptyView searchEmptyView;
    private WikipediaApp app;

    private String currentSearchQuery;
    private CompositeDisposable disposables = new CompositeDisposable();
    private HistoryEntryItemAdapter adapter = new HistoryEntryItemAdapter();

    private ItemCallback itemCallback = new ItemCallback();
    private ActionMode actionMode;
    private SearchActionModeCallback searchActionModeCallback = new HistorySearchCallback();
    private HashSet<HistoryEntry> selectedEntries = new HashSet<>();
    @NonNull public static HistoryFragment newInstance() {
        return new HistoryFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        app = WikipediaApp.getInstance();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_history, container, false);
        unbinder = ButterKnife.bind(this, view);

        searchEmptyView.setEmptyText(R.string.search_history_no_results);

        SwipeableItemTouchHelperCallback touchCallback = new SwipeableItemTouchHelperCallback(requireContext());
        touchCallback.setSwipeableEnabled(true);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(touchCallback);
        itemTouchHelper.attachToRecyclerView(historyList);

        historyList.setLayoutManager(new LinearLayoutManager(getContext()));
        historyList.setAdapter(adapter);
        historyEmptyView.setVisibility(View.GONE);

        setUpScrollListener();
        return view;
    }

    private void setUpScrollListener() {
        historyList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                ((MainActivity) requireActivity()).updateToolbarElevation(historyList.computeVerticalScrollOffset() != 0);
            }
        });
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onResume() {
        super.onResume();
        reloadHistoryItems();
    }

    @Override
    public void onPause() {
        super.onPause();
        if (actionMode != null) {
            actionMode.finish();
        }
    }

    @Override
    public void onDestroyView() {
        disposables.clear();
        historyList.setAdapter(null);
        historyList.clearOnScrollListeners();
        adapter.clearList();
        unbinder.unbind();
        unbinder = null;
        super.onDestroyView();
    }

    @Override
    public boolean onBackPressed() {
        if (actionMode != null) {
            actionMode.finish();
            return true;
        }
        if (selectedEntries.size() > 0) {
            unselectAllPages();
            return true;
        }
        return false;
    }

    private void updateEmptyState(@Nullable String searchQuery) {
        if (TextUtils.isEmpty(searchQuery)) {
            searchEmptyView.setVisibility(View.GONE);
            historyEmptyView.setVisibility(adapter.isEmpty() ? View.VISIBLE : View.GONE);
        } else {
            searchEmptyView.setVisibility(adapter.isEmpty() ? View.VISIBLE : View.GONE);
            historyEmptyView.setVisibility(View.GONE);
        }
    }

    private void onPageClick(@NonNull HistoryEntry entry) {
        Callback callback = callback();
        if (callback != null) {
            callback.onLoadPage(entry);
        }
    }

    private void onClearHistoryClick() {
        Callback callback = callback();
        if (callback != null) {
            callback.onClearHistory();
            reloadHistoryItems();
        }
    }

    private void finishActionMode() {
        if (actionMode != null) {
            actionMode.finish();
        }
    }

    private void beginMultiSelect() {
        if (HistorySearchCallback.is(actionMode)) {
            finishActionMode();
        }
    }

    private void toggleSelectPage(@Nullable IndexedHistoryEntry indexedEntry) {
        if (indexedEntry == null) {
            return;
        }
        if (selectedEntries.contains(indexedEntry.getEntry())) {
            selectedEntries.remove(indexedEntry.getEntry());
        } else {
            selectedEntries.add(indexedEntry.getEntry());
        }
        int selectedCount = selectedEntries.size();
        if (selectedCount == 0) {
            finishActionMode();
        } else if (actionMode != null) {
            actionMode.setTitle(getResources().getQuantityString(R.plurals.multi_items_selected, selectedCount, selectedCount));
        }
        adapter.notifyDataSetChanged();
    }

    public void refresh() {
        adapter.notifyDataSetChanged();
        if (!app.isOnline() && Prefs.shouldShowHistoryOfflineArticlesToast()) {
            Toast.makeText(requireContext(), R.string.history_offline_articles_toast, Toast.LENGTH_SHORT).show();
            Prefs.shouldShowHistoryOfflineArticlesToast(false);
        }
    }

    private void unselectAllPages() {
        selectedEntries.clear();
        adapter.notifyDataSetChanged();
    }

    private void deleteSelectedPages() {
        List<HistoryEntry> selectedEntryList = new ArrayList<>();
        for (HistoryEntry entry : selectedEntries) {
            if (entry != null) {
                selectedEntryList.add(entry);
                app.getDatabaseClient(HistoryEntry.class).delete(entry,
                        PageHistoryContract.PageWithImage.SELECTION);
            }
        }
        selectedEntries.clear();
        if (!selectedEntryList.isEmpty()) {
            showDeleteItemsUndoSnackbar(selectedEntryList);
            reloadHistoryItems();
        }
    }

    private void showDeleteItemsUndoSnackbar(final List<HistoryEntry> entries) {
        String message = entries.size() == 1
                ? getString(R.string.history_item_deleted, entries.get(0).getTitle().getDisplayText())
                : getString(R.string.history_items_deleted, entries.size());
        Snackbar snackbar = FeedbackUtil.makeSnackbar(getActivity(), message,
                FeedbackUtil.LENGTH_DEFAULT);
        snackbar.setAction(R.string.history_item_delete_undo, (v) -> {
            DatabaseClient<HistoryEntry> client = app.getDatabaseClient(HistoryEntry.class);
            for (HistoryEntry entry : entries) {
                client.upsert(entry, PageHistoryContract.PageWithImage.SELECTION);
            }
            reloadHistoryItems();
        });
        snackbar.show();
    }

    private void reloadHistoryItems() {
        disposables.clear();
        disposables.add(Observable.fromCallable(() -> HistoryDbHelper.INSTANCE.filterHistoryItems(StringUtils.defaultString(currentSearchQuery)))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onLoadItemsFinished, t -> {
                    L.e(t);
                    onLoadItemsFinished(Collections.emptyList());
                }));
    }

    private void onLoadItemsFinished(@NonNull List<Object> items) {
        List<Object> list = new ArrayList<>();
        if (!HistorySearchCallback.is(actionMode)) {
            list.add(new SearchBar());
        }
        list.addAll(items);
        adapter.setList(list);
        updateEmptyState(currentSearchQuery);
        requireActivity().invalidateOptionsMenu();
    }

    public static class IndexedHistoryEntry {
        @NonNull private final HistoryEntry entry;
        @Nullable private final String imageUrl;

        public IndexedHistoryEntry(@NonNull Cursor cursor) {
            this.entry = HistoryEntry.DATABASE_TABLE.fromCursor(cursor);
            this.imageUrl = PageHistoryContract.PageWithImage.IMAGE_NAME.val(cursor);
        }

        @Nullable public String getImageUrl() {
            return imageUrl;
        }

        @NonNull public HistoryEntry getEntry() {
            return entry;
        }
    }

    private static class HeaderViewHolder extends DefaultViewHolder<View> {
        TextView headerText;
        HeaderViewHolder(View itemView) {
            super(itemView);
            headerText = itemView.findViewById(R.id.section_header_text);
        }

        void bindItem(@NonNull String date) {
            headerText.setText(date);
        }
    }

    private class SearchCardViewHolder extends DefaultViewHolder<View> {
        private ImageView historyFilterButton;
        private ImageView clearHistoryButton;
        private WikiCardView searchWikiCardView;

        SearchCardViewHolder(View itemView) {
            super(itemView);
            searchWikiCardView = itemView.findViewById(R.id.wiki_card_for_search);
            View searchCardView = itemView.findViewById(R.id.search_card);
            AppCompatImageView voiceSearchButton = itemView.findViewById(R.id.voice_search_button);
            historyFilterButton = itemView.findViewById(R.id.history_filter);
            clearHistoryButton = itemView.findViewById(R.id.history_delete);
            searchCardView.setOnClickListener(view -> ((MainFragment) getParentFragment()).openSearchActivity(Constants.InvokeSource.NAV_MENU, null, searchWikiCardView));
            voiceSearchButton.setOnClickListener(view -> ((MainFragment) getParentFragment()).onFeedVoiceSearchRequested());
            historyFilterButton.setOnClickListener(view -> {
                if (actionMode == null) {
                    actionMode = ((AppCompatActivity) requireActivity())
                            .startSupportActionMode(searchActionModeCallback);
                }
            });
            clearHistoryButton.setOnClickListener(view -> {
                if (selectedEntries.size() == 0) {
                    new AlertDialog.Builder(requireContext())
                            .setTitle(R.string.dialog_title_clear_history)
                            .setMessage(R.string.dialog_message_clear_history)
                            .setPositiveButton(R.string.dialog_message_clear_history_yes, (dialog, which) -> onClearHistoryClick())
                            .setNegativeButton(R.string.dialog_message_clear_history_no, null).create().show();
                } else {
                    deleteSelectedPages();
                }
            });
            FeedbackUtil.setButtonLongPressToast(historyFilterButton, clearHistoryButton);
            searchWikiCardView.setCardBackgroundColor(ResourceUtil.getThemedColor(requireContext(), R.attr.color_group_22));
        }

        public void bindItem() {
            clearHistoryButton.setVisibility(adapter.isEmpty() ? View.GONE : View.VISIBLE);
            historyFilterButton.setVisibility(adapter.isEmpty() ? View.GONE : View.VISIBLE);
        }
    }

    private class HistoryEntryItemHolder extends DefaultViewHolder<PageItemView<IndexedHistoryEntry>>
            implements SwipeableItemTouchHelperCallback.Callback {
        private HistoryEntry entry;

        HistoryEntryItemHolder(PageItemView<IndexedHistoryEntry> itemView) {
            super(itemView);
        }

        void bindItem(@NonNull IndexedHistoryEntry indexedEntry) {
            entry = indexedEntry.getEntry();
            getView().setItem(indexedEntry);
            getView().setTitle(indexedEntry.getEntry().getTitle().getDisplayText());
            getView().setDescription(indexedEntry.getEntry().getTitle().getDescription());
            getView().setImageUrl(indexedEntry.getImageUrl());
            getView().setSelected(selectedEntries.contains(indexedEntry.getEntry()));
            PageAvailableOfflineHandler.INSTANCE.check(indexedEntry.getEntry().getTitle(), available -> getView().setViewsGreyedOut(!available));
        }


        @Override
        public void onSwipe() {
            selectedEntries.add(entry);
            deleteSelectedPages();
        }
    }

    private final class HistoryEntryItemAdapter extends RecyclerView.Adapter<DefaultViewHolder> {

        private static final int VIEW_TYPE_SEARCH_CARD = 0;
        private static final int VIEW_TYPE_HEADER = 1;
        private static final int VIEW_TYPE_ITEM = 2;

        @NonNull
        private List<Object> historyEntries = new ArrayList<>();

        @Override
        public int getItemCount() {
            return historyEntries.size();
        }

        public boolean isEmpty() {
            return (getItemCount() == 0)
                    || (getItemCount() == 1 && historyEntries.get(0) instanceof SearchBar);
        }

        @Override
        public int getItemViewType(int position) {
            if (historyEntries.get(position) instanceof SearchBar) {
                return VIEW_TYPE_SEARCH_CARD;
            } else if (historyEntries.get(position) instanceof String) {
                return VIEW_TYPE_HEADER;
            } else {
                return VIEW_TYPE_ITEM;
            }
        }

        public void setList(@NonNull List<Object> list) {
            historyEntries = list;
            this.notifyDataSetChanged();
        }

        void clearList() {
            historyEntries.clear();
        }

        @Override
        public DefaultViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            if (viewType == VIEW_TYPE_SEARCH_CARD) {
                View view = LayoutInflater.from(requireContext()).inflate(R.layout.view_history_header_with_search, parent, false);
                return new SearchCardViewHolder(view);
            } else if (viewType == VIEW_TYPE_HEADER) {
                View view = LayoutInflater.from(requireContext()).inflate(R.layout.view_section_header, parent, false);
                return new HeaderViewHolder(view);
            } else {
                return new HistoryEntryItemHolder(new PageItemView<>(requireContext()));
            }
        }

        @Override
        public void onBindViewHolder(@NonNull DefaultViewHolder holder, int pos) {
            if (holder instanceof SearchCardViewHolder) {
                ((SearchCardViewHolder) holder).bindItem();
            } else if (holder instanceof HistoryEntryItemHolder) {
                ((HistoryEntryItemHolder) holder).bindItem((IndexedHistoryEntry) historyEntries.get(pos));
            } else {
                ((HeaderViewHolder) holder).bindItem((String) historyEntries.get(pos));
            }
        }

        @Override public void onViewAttachedToWindow(@NonNull DefaultViewHolder holder) {
            super.onViewAttachedToWindow(holder);
            if (holder instanceof HistoryEntryItemHolder) {
                ((HistoryEntryItemHolder) holder).getView().setCallback(itemCallback);
            }
        }

        @Override public void onViewDetachedFromWindow(@NonNull DefaultViewHolder holder) {
            if (holder instanceof HistoryEntryItemHolder) {
                ((HistoryEntryItemHolder) holder).getView().setCallback(null);
            }
            super.onViewDetachedFromWindow(holder);
        }

        public void hideHeader() {
            if (!historyEntries.isEmpty() && historyEntries.get(0) instanceof SearchBar) {
                historyEntries.remove(0);
                notifyDataSetChanged();
            }
        }
    }

    private class ItemCallback implements PageItemView.Callback<IndexedHistoryEntry> {
        @Override
        public void onClick(@Nullable IndexedHistoryEntry indexedEntry) {
            if (selectedEntries != null && !selectedEntries.isEmpty()) {
                toggleSelectPage(indexedEntry);
            } else if (indexedEntry != null) {
                onPageClick(new HistoryEntry(indexedEntry.getEntry().getTitle(), HistoryEntry.SOURCE_HISTORY));
            }
        }

        @Override
        public boolean onLongClick(@Nullable IndexedHistoryEntry indexedEntry) {
            beginMultiSelect();
            toggleSelectPage(indexedEntry);
            return true;
        }

        @Override
        public void onThumbClick(@Nullable IndexedHistoryEntry indexedEntry) {
            onClick(indexedEntry);
        }

        @Override
        public void onActionClick(@Nullable IndexedHistoryEntry entry, @NonNull View view) {
        }
        @Override
        public void onSecondaryActionClick(@Nullable IndexedHistoryEntry entry, @NonNull View view) {
        }

        @Override
        public void onListChipClick(@NonNull ReadingList readingList) {
        }
    }

    private class HistorySearchCallback extends SearchActionModeCallback {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            actionMode = mode;
            ((MainFragment) getParentFragment()).setBottomNavVisible(false);
            ((HistoryEntryItemAdapter)historyList.getAdapter()).hideHeader();
            return super.onCreateActionMode(mode, menu);
        }

        @Override
        protected void onQueryChange(String s) {
            currentSearchQuery = s.trim();
            reloadHistoryItems();
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            super.onDestroyActionMode(mode);
            currentSearchQuery = "";
            reloadHistoryItems();
            actionMode = null;
            ((MainFragment) getParentFragment()).setBottomNavVisible(true);
        }

        @Override
        protected String getSearchHintString() {
            return requireContext().getResources().getString(R.string.history_filter_list_hint);
        }

        @Override
        protected Context getParentContext() {
            return requireContext();
        }
    }

    @Nullable private Callback callback() {
        return FragmentUtil.getCallback(this, Callback.class);
    }

    private static class SearchBar { }
}

File: app/src/main/java/org/wikipedia/history/HistoryFragment.kt
package org.wikipedia.history

import android.content.Context
import android.graphics.Typeface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.core.view.isVisible
import androidx.core.view.updateLayoutParams
import androidx.core.view.updateMarginsRelative
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import org.wikipedia.BackPressedHandler
import org.wikipedia.Constants
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.activity.FragmentUtil
import org.wikipedia.databinding.FragmentHistoryBinding
import org.wikipedia.main.MainActivity
import org.wikipedia.main.MainFragment
import org.wikipedia.page.PageAvailableOfflineHandler
import org.wikipedia.readinglist.database.ReadingList
import org.wikipedia.settings.Prefs
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.FeedbackUtil
import org.wikipedia.util.Resource
import org.wikipedia.util.ResourceUtil
import org.wikipedia.util.log.L
import org.wikipedia.views.DefaultViewHolder
import org.wikipedia.views.PageItemView
import org.wikipedia.views.SwipeableItemTouchHelperCallback
import org.wikipedia.views.WikiCardView

class HistoryFragment : Fragment(), BackPressedHandler {
    interface Callback {
        fun onLoadPage(entry: HistoryEntry)
    }

    private var _binding: FragmentHistoryBinding? = null
    private val binding get() = _binding!!
    private val viewModel: HistoryViewModel by viewModels()

    private val adapter = HistoryEntryItemAdapter()
    private val itemCallback = ItemCallback()
    private var actionMode: ActionMode? = null
    private val searchActionModeCallback = HistorySearchCallback()
    private val selectedEntries = mutableSetOf<HistoryEntry>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentHistoryBinding.inflate(inflater, container, false)

        binding.searchEmptyView.setEmptyText(R.string.search_history_no_results)
        val touchCallback = SwipeableItemTouchHelperCallback(requireContext())
        touchCallback.swipeableEnabled = true
        val itemTouchHelper = ItemTouchHelper(touchCallback)
        itemTouchHelper.attachToRecyclerView(binding.historyList)
        binding.historyList.layoutManager = LinearLayoutManager(context)
        binding.historyList.adapter = adapter
        binding.historyEmptyContainer.visibility = View.GONE
        setUpScrollListener()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.historyItems.observe(viewLifecycleOwner) {
            if (it is Resource.Success) {
                onLoadItemsFinished(it.data)
            } else if (it is Resource.Error) {
                onError(it.throwable)
            }
        }

        viewModel.deleteHistoryItemsAction.observe(viewLifecycleOwner) {
            if (it is Resource.Success) {
                onPagesDeleted()
            }
        }
    }

    private fun setUpScrollListener() {
        binding.historyList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                (requireActivity() as MainActivity).updateToolbarElevation(binding.historyList.computeVerticalScrollOffset() != 0)
            }
        })
    }

    override fun onResume() {
        super.onResume()
        viewModel.reloadHistoryItems()
    }

    override fun onPause() {
        super.onPause()
        actionMode?.finish()
    }

    override fun onDestroyView() {
        binding.historyList.adapter = null
        binding.historyList.clearOnScrollListeners()
        adapter.clearList()
        _binding = null
        super.onDestroyView()
    }

    override fun onBackPressed(): Boolean {
        actionMode?.run {
            finish()
            return@onBackPressed true
        }
        if (selectedEntries.size > 0) {
            unselectAllPages()
            return true
        }
        return false
    }

    private fun updateEmptyState(searchQuery: String?) {
        if (searchQuery.isNullOrEmpty()) {
            binding.searchEmptyView.visibility = View.GONE
            binding.historyEmptyContainer.visibility = if (adapter.isEmpty) View.VISIBLE else View.GONE
        } else {
            binding.searchEmptyView.visibility = if (adapter.isEmpty) View.VISIBLE else View.GONE
            binding.historyEmptyContainer.visibility = View.GONE
        }
    }

    private fun onPageClick(entry: HistoryEntry) {
        callback()?.onLoadPage(entry)
    }

    private fun finishActionMode() {
        actionMode?.finish()
    }

    private fun beginMultiSelect() {
        if (SearchActionModeCallback.matches(actionMode)) {
            finishActionMode()
        }
    }

    private fun toggleSelectPage(entry: HistoryEntry?) {
        if (entry == null) {
            return
        }
        if (selectedEntries.contains(entry)) {
            selectedEntries.remove(entry)
        } else {
            selectedEntries.add(entry)
        }
        val selectedCount = selectedEntries.size
        if (selectedCount == 0) {
            finishActionMode()
        } else {
            actionMode?.title = resources.getQuantityString(R.plurals.multi_items_selected, selectedCount, selectedCount)
        }
        val position = adapter.getPosition(entry)
        adapter.notifyItemChanged(position)
    }

    fun refresh() {
        adapter.notifyItemRangeChanged(0, adapter.itemCount)
        if (!WikipediaApp.instance.isOnline && Prefs.showHistoryOfflineArticlesToast) {
            Toast.makeText(requireContext(), R.string.history_offline_articles_toast, Toast.LENGTH_SHORT).show()
            Prefs.showHistoryOfflineArticlesToast = false
        }
    }

    private fun unselectAllPages() {
        selectedEntries.clear()
        adapter.notifyItemRangeChanged(0, adapter.itemCount)
    }

    private fun deleteSelectedPages() {
        viewModel.deleteHistoryItems(selectedEntries.toList())
    }

    private fun onPagesDeleted() {
        showDeleteItemsUndoSnackbar(selectedEntries.toList())
        selectedEntries.clear()
        viewModel.reloadHistoryItems()
    }

    private fun showDeleteItemsUndoSnackbar(entries: List<HistoryEntry>) {
        val message = if (entries.size == 1) getString(R.string.history_item_deleted, entries[0].title.displayText) else getString(R.string.history_items_deleted, entries.size)
        val snackbar = FeedbackUtil.makeSnackbar(requireActivity(), message)
        snackbar.setAction(R.string.history_item_delete_undo) {
            viewModel.insertHistoryItem(entries)
        }
        snackbar.show()
    }

    private fun onLoadItemsFinished(items: List<Any>) {
        val list = mutableListOf<Any>()
        if (!SearchActionModeCallback.matches(actionMode)) {
            list.add(SearchBar())
        }
        list.addAll(items)
        adapter.setList(list)
        updateEmptyState(viewModel.searchQuery)
        requireActivity().invalidateOptionsMenu()
    }

    private fun onError(throwable: Throwable) {
        L.e(throwable)
        onLoadItemsFinished(emptyList())
    }

    private class HeaderViewHolder(itemView: View) : DefaultViewHolder<View>(itemView) {
        var headerText = itemView.findViewById<TextView>(R.id.section_header_text)!!

        fun bindItem(date: String) {
            headerText.text = date
        }
    }

    private inner class SearchCardViewHolder(itemView: View) : DefaultViewHolder<View>(itemView) {
        private val historyFilterButton: ImageView
        private val clearHistoryButton: ImageView

        fun bindItem() {
            clearHistoryButton.visibility = if (adapter.isEmpty) View.GONE else View.VISIBLE
            historyFilterButton.visibility = if (adapter.isEmpty) View.GONE else View.VISIBLE
        }

        private fun adjustSearchCardView(searchCardView: WikiCardView) {
            searchCardView.post {
                if (!isAdded) {
                    return@post
                }
                searchCardView.updateLayoutParams<LinearLayout.LayoutParams> {
                    val horizontalMargin = if (DimenUtil.isLandscape(requireContext())) {
                        searchCardView.width / 6 + DimenUtil.roundedDpToPx(30f)
                    } else {
                        DimenUtil.roundedDpToPx(16f)
                    }
                    updateMarginsRelative(start = horizontalMargin, end = horizontalMargin,
                        top = DimenUtil.roundedDpToPx(3f))
                }
            }
            searchCardView.setCardBackgroundColor(ResourceUtil.getThemedColor(requireContext(), R.attr.background_color))
        }

        init {
            val searchCardView = itemView.findViewById<WikiCardView>(R.id.search_card)
            val voiceSearchButton = itemView.findViewById<View>(R.id.voice_search_button)
            historyFilterButton = itemView.findViewById(R.id.history_filter)
            clearHistoryButton = itemView.findViewById(R.id.history_delete)
            searchCardView.setOnClickListener { (requireParentFragment() as MainFragment).openSearchActivity(Constants.InvokeSource.NAV_MENU, null, it) }
            voiceSearchButton.isVisible = WikipediaApp.instance.voiceRecognitionAvailable
            voiceSearchButton.setOnClickListener { (requireParentFragment() as MainFragment).onFeedVoiceSearchRequested() }
            historyFilterButton.setOnClickListener {
                if (actionMode == null) {
                    actionMode = (requireActivity() as AppCompatActivity)
                            .startSupportActionMode(searchActionModeCallback)
                }
            }
            clearHistoryButton.setOnClickListener {
                if (selectedEntries.size == 0) {
                    MaterialAlertDialogBuilder(requireContext())
                            .setTitle(R.string.dialog_title_clear_history)
                            .setMessage(R.string.dialog_message_clear_history)
                            .setPositiveButton(R.string.dialog_message_clear_history_yes) { _, _ -> viewModel.deleteAllHistoryItems() }
                            .setNegativeButton(R.string.dialog_message_clear_history_no, null).show()
                } else {
                    deleteSelectedPages()
                }
            }
            FeedbackUtil.setButtonTooltip(historyFilterButton, clearHistoryButton)
            adjustSearchCardView(searchCardView)
        }
    }

    private inner class HistoryEntryItemHolder(itemView: PageItemView<HistoryEntry>) : DefaultViewHolder<PageItemView<HistoryEntry>>(itemView), SwipeableItemTouchHelperCallback.Callback {
        private lateinit var entry: HistoryEntry

        fun bindItem(entry: HistoryEntry) {
            this.entry = entry
            view.item = entry
            view.setTitle(entry.title.displayText)
            view.setTitleTypeface(Typeface.NORMAL)
            view.setDescription(entry.title.description)
            view.setImageUrl(entry.title.thumbUrl)
            view.isSelected = selectedEntries.contains(entry)
            PageAvailableOfflineHandler.check(lifecycleScope, entry.title) { view.setViewsGreyedOut(!it) }
        }

        override fun onSwipe() {
            selectedEntries.add(entry)
            deleteSelectedPages()
        }

        override fun isSwipeable(): Boolean { return true }
    }

    private inner class HistoryEntryItemAdapter : RecyclerView.Adapter<DefaultViewHolder<*>>() {
        private var historyEntries = mutableListOf<Any>()
        override fun getItemCount(): Int {
            return historyEntries.size
        }

        val isEmpty
            get() = (itemCount == 0 || itemCount == 1 && historyEntries[0] is SearchBar)

        override fun getItemViewType(position: Int): Int {
            return when {
                historyEntries[position] is SearchBar -> Companion.VIEW_TYPE_SEARCH_CARD
                historyEntries[position] is String -> Companion.VIEW_TYPE_HEADER
                else -> Companion.VIEW_TYPE_ITEM
            }
        }

        fun setList(list: MutableList<Any>) {
            historyEntries = list
            adapter.notifyDataSetChanged()
        }

        fun clearList() {
            historyEntries.clear()
        }

        fun getPosition(entry: HistoryEntry): Int {
            return historyEntries.indexOf(entry)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DefaultViewHolder<*> {
            return when (viewType) {
                Companion.VIEW_TYPE_SEARCH_CARD -> {
                    val view = LayoutInflater.from(requireContext()).inflate(R.layout.view_history_header_with_search, parent, false)
                    SearchCardViewHolder(view)
                }
                Companion.VIEW_TYPE_HEADER -> {
                    val view = LayoutInflater.from(requireContext()).inflate(R.layout.view_section_header, parent, false)
                    HeaderViewHolder(view)
                }
                else -> HistoryEntryItemHolder(PageItemView(requireContext()))
            }
        }

        override fun onBindViewHolder(holder: DefaultViewHolder<*>, pos: Int) {
            when (holder) {
                is SearchCardViewHolder -> holder.bindItem()
                is HistoryEntryItemHolder -> holder.bindItem(historyEntries[pos] as HistoryEntry)
                else -> (holder as HeaderViewHolder).bindItem(historyEntries[pos] as String)
            }
        }

        override fun onViewAttachedToWindow(holder: DefaultViewHolder<*>) {
            super.onViewAttachedToWindow(holder)
            if (holder is HistoryEntryItemHolder) {
                holder.view.callback = itemCallback
            }
        }

        override fun onViewDetachedFromWindow(holder: DefaultViewHolder<*>) {
            if (holder is HistoryEntryItemHolder) {
                holder.view.callback = null
            }
            super.onViewDetachedFromWindow(holder)
        }

        fun hideHeader() {
            if (historyEntries.isNotEmpty() && historyEntries[0] is SearchBar) {
                historyEntries.removeAt(0)
                adapter.notifyItemRemoved(0)
            }
        }
    }

    private inner class ItemCallback : PageItemView.Callback<HistoryEntry?> {
        override fun onClick(item: HistoryEntry?) {
            if (selectedEntries.isNotEmpty()) {
                toggleSelectPage(item)
            } else if (item != null) {
                onPageClick(HistoryEntry(item.title, HistoryEntry.SOURCE_HISTORY))
            }
        }

        override fun onLongClick(item: HistoryEntry?): Boolean {
            beginMultiSelect()
            toggleSelectPage(item)
            return true
        }

        override fun onActionClick(item: HistoryEntry?, view: View) {}
        override fun onListChipClick(readingList: ReadingList) {}
    }

    private inner class HistorySearchCallback : SearchActionModeCallback() {
        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            actionMode = mode
            (requireParentFragment() as MainFragment).setBottomNavVisible(false)
            (binding.historyList.adapter as HistoryEntryItemAdapter).hideHeader()
            return super.onCreateActionMode(mode, menu)
        }

        override fun onQueryChange(s: String) {
            viewModel.searchQuery = s.trim()
            viewModel.reloadHistoryItems()
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            super.onDestroyActionMode(mode)
            viewModel.searchQuery = ""
            viewModel.reloadHistoryItems()
            actionMode = null
            (requireParentFragment() as MainFragment).setBottomNavVisible(true)
        }

        override fun getSearchHintString(): String {
            return requireContext().resources.getString(R.string.history_filter_list_hint)
        }

        override fun getParentContext(): Context {
            return requireContext()
        }
    }

    private fun callback(): Callback? {
        return FragmentUtil.getCallback(this, Callback::class.java)
    }

    private class SearchBar

    companion object {
        private const val VIEW_TYPE_SEARCH_CARD = 0
        private const val VIEW_TYPE_HEADER = 1
        private const val VIEW_TYPE_ITEM = 2

        fun newInstance(): HistoryFragment {
            return HistoryFragment()
        }
    }
}
